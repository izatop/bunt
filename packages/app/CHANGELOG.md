# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

# [0.7.0](https://github.com/izatop/typesafeunit/compare/v0.6.28...v0.7.0) (2020-05-26)


### Features

* new eslint rules ([62417eb](https://github.com/izatop/typesafeunit/commit/62417eb1385862c93108991075f864536b90a748))





## [0.6.28](https://github.com/izatop/typesafeunit/compare/v0.6.27...v0.6.28) (2020-05-17)

**Note:** Version bump only for package @typesafeunit/app





## [0.6.27](https://github.com/izatop/typesafeunit/compare/v0.6.26...v0.6.27) (2020-05-15)

**Note:** Version bump only for package @typesafeunit/app





## [0.6.26](https://github.com/izatop/typesafeunit/compare/v0.6.25...v0.6.26) (2020-05-12)

**Note:** Version bump only for package @typesafeunit/app





## [0.6.22](https://github.com/izatop/typesafeunit/compare/v0.6.21...v0.6.22) (2020-05-07)

**Note:** Version bump only for package @typesafeunit/app





## [0.6.21](https://github.com/izatop/typesafeunit/compare/v0.6.20...v0.6.21) (2020-05-06)

**Note:** Version bump only for package @typesafeunit/app





## [0.6.20](https://github.com/izatop/typesafeunit/compare/v0.6.19...v0.6.20) (2020-05-06)

**Note:** Version bump only for package @typesafeunit/app





## [0.6.19](https://github.com/izatop/typesafeunit/compare/v0.6.18...v0.6.19) (2020-05-01)

**Note:** Version bump only for package @typesafeunit/app





## [0.6.18](https://github.com/izatop/typesafeunit/compare/v0.6.17...v0.6.18) (2020-04-30)


### Bug Fixes

* logging ([2365a0b](https://github.com/izatop/typesafeunit/commit/2365a0b12902300b2e898e2c07aa5769be3b8e56))





## [0.6.15](https://github.com/izatop/typesafeunit/compare/v0.6.14...v0.6.15) (2020-04-04)


### Bug Fixes

* action state ([32974d7](https://github.com/izatop/typesafeunit/commit/32974d7ae0723888531d376d2642d9d62e0954f8))





## [0.6.14](https://github.com/izatop/typesafeunit/compare/v0.6.13...v0.6.14) (2020-04-03)

**Note:** Version bump only for package @typesafeunit/app





## [0.6.13](https://github.com/izatop/typesafeunit/compare/v0.6.12...v0.6.13) (2020-04-01)


### Bug Fixes

* runtime, tests ([c17980f](https://github.com/izatop/typesafeunit/commit/c17980fb75e047dfc001e6fca917ef1f3a9d84dd))





## [0.6.12](https://github.com/izatop/typesafeunit/compare/v0.6.11...v0.6.12) (2020-04-01)

**Note:** Version bump only for package @typesafeunit/app





## [0.6.11](https://github.com/izatop/typesafeunit/compare/v0.6.10...v0.6.11) (2020-04-01)


### Bug Fixes

* action return type ([fe2dc96](https://github.com/izatop/typesafeunit/commit/fe2dc964e3e0776f15288f511cf7962cbf3fa8a3))





## [0.6.10](https://github.com/izatop/typesafeunit/compare/v0.6.9...v0.6.10) (2020-03-30)

**Note:** Version bump only for package @typesafeunit/app





## [0.6.9](https://github.com/izatop/typesafeunit/compare/v0.6.8...v0.6.9) (2020-03-30)

**Note:** Version bump only for package @typesafeunit/app





## [0.6.8](https://github.com/izatop/typesafeunit/compare/v0.6.7...v0.6.8) (2020-03-30)


### Bug Fixes

* errors ([5c61c09](https://github.com/izatop/typesafeunit/commit/5c61c09f19ddb0415238d8cd48681b9dddd74f29))





## [0.6.7](https://github.com/izatop/typesafeunit/compare/v0.6.6...v0.6.7) (2020-03-30)

**Note:** Version bump only for package @typesafeunit/app





## [0.6.6](https://github.com/izatop/typesafeunit/compare/v0.6.5...v0.6.6) (2020-03-30)

**Note:** Version bump only for package @typesafeunit/app





## [0.6.5](https://github.com/izatop/typesafeunit/compare/v0.6.4...v0.6.5) (2020-03-30)

**Note:** Version bump only for package @typesafeunit/app





## [0.6.4](https://github.com/izatop/typesafeunit/compare/v0.6.3...v0.6.4) (2020-03-30)

**Note:** Version bump only for package @typesafeunit/app





## [0.6.3](https://github.com/izatop/typesafeunit/compare/v0.6.2...v0.6.3) (2020-03-30)

**Note:** Version bump only for package @typesafeunit/app





## [0.6.2](https://github.com/izatop/typesafeunit/compare/v0.6.1...v0.6.2) (2020-03-30)

**Note:** Version bump only for package @typesafeunit/app





## [0.6.1](https://github.com/izatop/typesafeunit/compare/v0.6.0...v0.6.1) (2020-03-30)

**Note:** Version bump only for package @typesafeunit/app





# [0.6.0](https://github.com/izatop/typesafeunit/compare/v0.5.2...v0.6.0) (2020-03-30)


### Features

* logger, perf etc ([daed7fe](https://github.com/izatop/typesafeunit/commit/daed7fe86ea6e9cb3bd90830a77167e6f9e21755))





## [0.5.2](https://github.com/izatop/typesafeunit/compare/v0.5.1...v0.5.2) (2020-03-26)

**Note:** Version bump only for package @typesafeunit/app





## [0.5.1](https://github.com/izatop/typesafeunit/compare/v0.5.0...v0.5.1) (2020-03-26)

**Note:** Version bump only for package @typesafeunit/app





# [0.5.0](https://github.com/izatop/typesafeunit/compare/v0.4.6...v0.5.0) (2020-03-26)

**Note:** Version bump only for package @typesafeunit/app





## [0.4.5](https://github.com/izatop/typesafeunit/compare/v0.4.4...v0.4.5) (2020-03-25)


### Bug Fixes

* improve application factory ([0899aa1](https://github.com/izatop/typesafeunit/commit/0899aa1329fb189812af65af6580cc228145d75c))





## [0.4.3](https://github.com/izatop/typesafeunit/compare/v0.4.2...v0.4.3) (2020-03-25)

**Note:** Version bump only for package @typesafeunit/app





## [0.4.2](https://github.com/izatop/typesafeunit/compare/v0.4.1...v0.4.2) (2020-03-25)

**Note:** Version bump only for package @typesafeunit/app





## [0.4.1](https://github.com/izatop/typesafeunit/compare/v0.4.0...v0.4.1) (2020-03-25)

**Note:** Version bump only for package @typesafeunit/app





# [0.4.0](https://github.com/izatop/typesafeunit/compare/v0.3.9...v0.4.0) (2020-03-25)


### Features

* http server ([62abc0f](https://github.com/izatop/typesafeunit/commit/62abc0f20cfefac0a7d5fa6bc91c030719d5e9a8))





## [0.3.9](https://github.com/izatop/typesafeunit/compare/v0.3.8...v0.3.9) (2020-03-08)

**Note:** Version bump only for package @typesafeunit/app





## [0.3.8](https://github.com/izatop/typesafeunit/compare/v0.3.7...v0.3.8) (2020-03-08)

**Note:** Version bump only for package @typesafeunit/app





## [0.3.7](https://github.com/izatop/typesafeunit/compare/v0.3.6...v0.3.7) (2020-03-08)

**Note:** Version bump only for package @typesafeunit/app





## [0.3.6](https://github.com/izatop/typesafeunit/compare/v0.3.5...v0.3.6) (2020-03-08)


### Bug Fixes

* error handling ([1792e2b](https://github.com/izatop/typesafeunit/commit/1792e2b83fec230a396d955ab12ccea96e7bc758))





## [0.3.5](https://github.com/izatop/typesafeunit/compare/v0.3.4...v0.3.5) (2020-03-07)


### Bug Fixes

* headers keys case sensitivity ([37c372d](https://github.com/izatop/typesafeunit/commit/37c372d23a1ac9a5175aa6428c3b1c82e17e669c))
* **headers:** case sensitivity fix ([c119eae](https://github.com/izatop/typesafeunit/commit/c119eae6fb96fba11f27c260f130b351addc34fa))





## [0.3.4](https://github.com/izatop/typesafeunit/compare/v0.3.3...v0.3.4) (2020-03-07)


### Bug Fixes

* type fix ([a41f766](https://github.com/izatop/typesafeunit/commit/a41f76606330f83a8355081fdcc7d9c112708efe))





## [0.3.3](https://github.com/izatop/typesafeunit/compare/v0.3.2...v0.3.3) (2020-03-05)


### Bug Fixes

* **npm:** add LICENSE ([d7bf57f](https://github.com/izatop/typesafeunit/commit/d7bf57fe2dcb9f83cc4a570c99fafcb5e318bdd4))





## [0.3.2](https://github.com/izatop/typesafeunit/compare/v0.3.1...v0.3.2) (2020-03-05)

**Note:** Version bump only for package @typesafeunit/app





## [0.3.1](https://github.com/izatop/typesafeunit/compare/v0.3.0...v0.3.1) (2020-03-05)

**Note:** Version bump only for package @typesafeunit/app





# 0.3.0 (2020-03-05)


### Features

* initial commit ([a76a484](https://github.com/izatop/typesafeunit/commit/a76a4849282c382311f4d0cb9f9bd66a10bcceaa))
