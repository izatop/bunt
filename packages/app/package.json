{
    "name": "@typesafeunit/app",
    "version": "0.6.12",
    "keywords": [
        "typescript"
    ],
    "author": {
        "name": "Artur Bier",
        "email": "izatop@gmail.com"
    },
    "main": "dist/index.js",
    "types": "dist/index.d.ts",
    "files": [
        "dist/"
    ],
    "description": "Application Server",
    "repository": "git@github.com:izatop/typesafeunit.git",
    "scripts": {
        "clean": "yarn build:clean",
        "watch": "yarn build:watch",
        "build": "yarn build:clean && tsc",
        "build:clean": "rimraf dist tsconfig.tsbuildinfo",
        "build:watch": "yarn build --watch"
    },
    "publishConfig": {
        "access": "public"
    },
    "dependencies": {
        "@typesafeunit/unit": "^0.6.12",
        "@typesafeunit/util": "^0.6.12",
        "path-to-regexp": "^6.1.0"
    },
    "license": "MIT"
}
