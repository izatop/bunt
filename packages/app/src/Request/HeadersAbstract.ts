import {assert, ILogable, isBoolean, isFunction} from "@bunt/util";
import {HeaderAssertValue, IHeaders} from "../interfaces.js";
import {StrictKeyValueMap} from "./StrictKeyValueMap.js";

export abstract class HeadersAbstract extends StrictKeyValueMap
    implements IHeaders, ILogable<{[key: string]: string}> {

    public assert(header: string, expected: HeaderAssertValue): void {
        const clientValue = this.get(header.toLowerCase());
        if (Array.isArray(expected)) {
            assert(
                expected.some((e) => clientValue.includes(e)),
                `Wrong header "${header}" value, allowed: ${expected.join(", ")}`,
            );
        }

        if (isFunction(expected)) {
            const result = expected(clientValue);
            if (isBoolean(result)) {
                assert(result, `Wrong ${header}`);
            }
        }
    }

    public getLogValue(): Record<any, any> {
        return this.toJSON();
    }
}
