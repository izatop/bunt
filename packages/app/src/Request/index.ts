export * from "./KeyValueMap.js";
export * from "./HeadersAbstract.js";
export * from "./RequestAbstract.js";
export * from "./RequestMessageAbstract.js";
export * from "./RequestValidatorAbstract.js";
export * from "./StrictKeyValueMap.js";
