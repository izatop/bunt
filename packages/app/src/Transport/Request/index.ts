export * from "./URLEncodedFormTransform";
export * from "./JSONTransform";
