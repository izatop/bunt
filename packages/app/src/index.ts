export * from "./Payload/index.js";
export * from "./Route/index.js";
export * from "./Request/index.js";
export * from "./TransformRequest/index.js";
export * from "./Application.js";
export * from "./interfaces.js";
