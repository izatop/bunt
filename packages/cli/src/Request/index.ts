export * from "./Headers.js";
export * from "./RequestCommand.js";
