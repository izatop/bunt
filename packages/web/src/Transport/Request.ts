import {IHeaders, KeyValueMap, RequestAbstract, RouteResponse} from "@typesafeunit/app";
import {
    isBoolean,
    isError,
    isNull,
    isNumber,
    isObject,
    isReadableStream,
    isString,
    isUndefined,
} from "@typesafeunit/util";
import {IncomingMessage, ServerResponse} from "http";
import {parse} from "url";
import {Headers} from "./Headers";
import {ResponseAbstract} from "./Response";
import {TransformError} from "./TransformError";

interface ISendOptions {
    code: number;
    status?: string;
    headers?: { [key: string]: string };
}

export class Request extends RequestAbstract {
    public readonly headers: IHeaders;
    public readonly route: string;

    private readonly incomingMessage: IncomingMessage;
    private readonly serverResponse: ServerResponse;

    constructor(incomingMessage: IncomingMessage, serverResponse: ServerResponse) {
        super();
        this.incomingMessage = incomingMessage;
        this.serverResponse = serverResponse;

        const headers: [string, string][] = [];
        for (const [key, value] of Object.entries(this.incomingMessage.headers)) {
            if (isString(value)) {
                headers.push([key, value]);
            }
        }

        this.route = this.getRoute();
        this.headers = new Headers(headers);
    }

    public createReadableStream(): NodeJS.ReadableStream {
        return this.incomingMessage;
    }

    protected getRoute() {
        const {pathname} = parse(this.incomingMessage.url || "/", true);
        const {method = "GET"} = this.incomingMessage;
        return `${method} ${pathname}`;
    }

    /**
     * @param response
     */
    protected async write(response: RouteResponse) {
        // @TODO Validate types in the Accept header before send a response.

        if (isUndefined(response) || isNull(response)) {
            return this.send("");
        }

        if (isString(response) || isNumber(response) || isBoolean(response)) {
            return this.send(response.toString());
        }

        if (isObject(response)) {
            if (Buffer.isBuffer(response)) {
                return this.send(response);
            }

            if (isReadableStream(response)) {
                response.pipe(this.serverResponse);
                return;
            }

            if (isError(response)) {
                const transform = new TransformError(response);
                const accept = this.headers.get("accept");
                const {response: transformed, ...status} = accept.includes("application/json")
                    ? transform.toJSON()
                    : transform.toString();

                if (accept.includes("application/json")) {
                    return this.send(transformed, {...status, headers: {"content-type": "application/json"}});
                }

                return this.send(transformed, {...status});
            }

            if (response instanceof ResponseAbstract) {
                const {code, status} = response;
                return this.send(
                    response.stringify(),
                    {code, status, headers: {"content-type": response.getContentType()}},
                );
            }

            return this.send(JSON.stringify(response));
        }
    }

    /**
     * @param body
     * @param options
     */
    protected send(body: string | undefined | Buffer, options: ISendOptions = {code: 200}) {
        try {
            const {code, status} = options;
            const headers = new KeyValueMap(Object.entries(options.headers || {}));
            if (!headers.has("content-type")) {
                headers.set("context-type", "text/plain; charset=utf-8");
            }

            for (const [header, value] of headers.entries()) {
                this.serverResponse.setHeader(header, value);
            }

            this.serverResponse.writeHead(code, status);
            this.serverResponse.write(body);
        } finally {
            this.serverResponse.end();
        }
    }
}
