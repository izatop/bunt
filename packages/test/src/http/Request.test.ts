import {JSONTransform, URLEncodedFormTransform} from "@typesafeunit/app";
import {Request} from "./src/transport/Request";

describe("Request", () => {
    test("JSONTransform", async () => {
        const request = new Request(
            "/u/123",
            {"Content-Type": "application/json"},
            JSON.stringify({field1: "Value1", num: 123}),
        );

        expect(await JSONTransform(request))
            .toMatchSnapshot();
    });

    test("FormTransform", async () => {
        const request = new Request(
            "/u/123",
            {"Content-Type": "application/json"},
            "key=value&foo=bar",
        );

        expect(await URLEncodedFormTransform(request))
            .toMatchSnapshot();
    });
});
