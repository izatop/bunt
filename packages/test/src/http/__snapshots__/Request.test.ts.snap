// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Request FormTransform 1`] = `
Object {
  "foo": "bar",
  "key": "value",
}
`;

exports[`Request JSONTransform 1`] = `
Object {
  "field1": "Value1",
  "num": 123,
}
`;
